h2o-automl/src/main/java/ai/h2o/automl/AutoML.java

```diff
@@ -164,76 +166,142 @@ public AutoML() {
 
   public AutoML(Key<AutoML> key, Date startTime, AutoMLBuildSpec buildSpec) {
     super(key == null ? Key.make(buildSpec.project()) : key);
-    _startTime = startTime;
-    _buildSpec = buildSpec;
-    _runCountdown = Countdown.fromSeconds(buildSpec.build_control.stopping_criteria.max_runtime_secs());
-    _modelingPlan = buildSpec.build_models.modeling_plan == null ? defaultModelingPlan : buildSpec.build_models.modeling_plan;
-    _modelingStepsRegistry = new ModelingStepsRegistry();
-
     try {
+      _startTime = startTime;
       _eventLog = EventLog.getOrMake(_key);
-      eventLog().info(Stage.Workflow, "Project: " + projectName());
-
-      handleCVParameters(buildSpec);
-
-      handleReproducibilityParameters(buildSpec);
-
-      handleDatafileParameters(buildSpec);
+      eventLog().info(Stage.Workflow, "Project: "+buildSpec.project());
 
-      handleEarlyStoppingParameters(buildSpec);
-
-      initLeaderboard(buildSpec);
+      validateBuildSpec(buildSpec);
+      _buildSpec = buildSpec;
+      // now that buildSpec is validated, we can assign it: all future logic can now safely access parameters through _buildSpec.
+      _runCountdown = Countdown.fromSeconds(_buildSpec.build_control.stopping_criteria.max_runtime_secs());
 
+      prepareData();
+      initLeaderboard();
       planWork();
-
       _modelingStepsExecutor = new ModelingStepsExecutor(_leaderboard, _eventLog, _runCountdown);
     } catch (Exception e) {
       delete(); //cleanup potentially leaked keys
       throw e;
     }
   }
 
-  private void initLeaderboard(AutoMLBuildSpec buildSpec) {
-    String sort_metric = buildSpec.input_spec.sort_metric;
-    sort_metric = sort_metric == null || StoppingMetric.AUTO.name().equalsIgnoreCase(sort_metric) ? null : sort_metric.toLowerCase();
-    _leaderboard = Leaderboard.getOrMake(_key.toString(), _eventLog, _leaderboardFrame, sort_metric);
-  }
+  private void validateBuildSpec(AutoMLBuildSpec buildSpec) {
+    // Validate all buildSpec parameters here and provide reasonable defaults or parameter cleaning if necessary.
+    // When assigning default or modifying the original value, use `eventLog` to inform the user about the changes.
+    // Please no additional logic!
```

Comment:
@seb-h2o "Please no additional logic!" - this IMO is redundant. It might be confusing if we will need to add extra validation for buildSpec

Reply:
yeah, bad wording, will fix.